openssl/sha.go

```diff
@@ -358,7 +407,10 @@ func (h *sha384Hash) MarshalBinary() ([]byte, error) {
 }
 
 func (h *sha512Hash) MarshalBinary() ([]byte, error) {
-	d := (*sha512Ctx)(unsafe.Pointer(&h.ctx))
+	d := (*sha512Ctx)(h.shaState())
```

Comment:
Maybe a silly question, but why do these implement `BinaryMarshaler`/`BinaryUnmarshaler`? Is roundtripping expected to work with OpenSSL-backed crypto vs. Boring vs. Go standard lib?

This goes all the way back to boring, I think, but seems worth asking about. 😄

Reply:
Implementing the marshaler interfaces is really not required by the [hash.Hash](https://pkg.go.dev/hash#Hash) interface, but all hash implementations in the standard library do implement them and this promise is covered by the Go compatibility promise:

> Compatibility: Any future changes to hash or crypto packages will endeavor to maintain compatibility with state encoded using previous versions. That is, any released versions of the packages should be able to decode data written with any previously released version, subject to issues such as security fixes.

-- Taken from  [hash.Hash](https://pkg.go.dev/hash#Hash).

Additionally, there are a bunch of tests in the Go repo that initialize hashes using their marshaled representation instead of the origin input data.

As we are replacing standard library hashes here, and it is documented that , we must also implement the marshalers interfaces,